Add multi-channel SDF font generation and rendering support.

bruvzg opened the pull request:
Support for M(T)SDF font rendering. Font texture is generated using https://github.com/Chlumsky/msdfgen.

<details>
  <summary>Zoom video</summary>

https://user-images.githubusercontent.com/7645683/109620721-debe3f00-7b42-11eb-870b-c2207812a03f.mov

</details>

<img width="1112" alt="Screenshot 2021-03-02 at 12 06 49" src="https://user-images.githubusercontent.com/7645683/109632554-d28cae80-7b4f-11eb-83c7-35aeaf87a0ab.png">

*Logos on background are rendered from 128x128 custom MSDF image.*

- Added caching for multiple font variations.
- Added font importing, with character range pre-rendering option, font are imported to `.fontdata` (and set of textures) which store all its properties (see file description below). It's still possible to load any of `.ttf` / `.otf` / `.font` / `.fnt` files directly.
- Adds option to convert dynamic fonts to bitmap (removes source font file from import, which will disable advanced shaping and OpenType features for the font, and limits it to pre-rendered glyphs only).
- Adds option to override global oversampling value (set to 0 to restore default behavior).
- Adds M(T)SDF glyph and outline font support using combined multi-channel and true signed distance textures:
![Screenshot 2021-02-25 at 15 14 05](https://user-images.githubusercontent.com/7645683/109277471-0ef29e80-7820-11eb-9b00-13ad2aeaa712.png)

### Import options:

`antialiased` - ignored for MSDF and bitmap fonts, if selected uses 8-bit greyscale rendering, otherwise uses 1-bit rendering.
`force_autohinter` - if selected prefers FreeType autohinter over font built-in hinter.
`msdf` - enables MSDF generation for dynamic fonts (or indicates that source is MSDF for bitmap fonts). MSDF generation is slow, but allows to use same texture for all fonts sizes, with almost no extra rendering cost.
`msdf_px_range` - MSDF pixel range (range around the shape between the minimum and maximum representable distance).
`oversampling` - font oversampling factor, ignored for MSDF and bitmap, if set to 0 global value is used.
`hintinhg` - ignored for bitmap fonts.
`base_variation` - default font size and variation, comma separated list of "size", "outline_size" or OpenType variation tags.
`convert_to_bitmap` - ignored for bitmap fonts, removes dynamic font source file form import (disables OpenType features, and limits font to pre-rendered glyphs only).
`extra_spacing_*` - extra spacing for glyphs and spaces.
`preload\ranges` - list of character ranges to preload in `'A'-'Z'`, `U+0000-U+0000` or `0x0000-0x0000` format, or glyph ranges in `G+0000-G+0000` format.
`support_overrides` - list of overrides of language and scripts, can be used to increase or decrease font fallback usage priority for specific language or script.

### Supported import variants:

TrueType / OpenType source -> pre-renders glyphs and outlines for each size, original font file is preserved in cache and used for OpenType features, shaping and rendering missing glyphs.
TrueType / OpenType source + `msdf` selected -> Same as above but pre-renders glyphs as MSDF for one size only. Outlines are generated from the same MSDF. Generating MSDF require outline and will not work with bitmap fonts (e.g. emoji fonts).
TrueType / OpenType source + `convert_to_bitmap` selected -> Same as above, removes original font file from cache (font have no OpenType features and limited to pre-rendered glyphs only).
AngelCode bitmap font -> font is copied to cache as is (`msdf` flag should be set manually, if source image is MSDF or M(T)SDF).

### ".fontdata" file format:

```
==============================================================================================================================================
type                         name                         description / values
==============================================================================================================================================
U32                          magic                        ASCII "GDFT"
U8                           version_major                1
U8                           version_minor                0
U32                          font_type                    ASCII "BMP0" - bitmap font, "DYN0" - dynamic font
U32                          font_data_size               size of dynamic font data in memory (always 0 for bitmap fonts)
if font_data_size > 0:
    U32                      font_comp_size               size of dynamic font data in file
    U8[font_comp_size]       font_data                    ZSTD compressed dynamic font data (source ttf/otf file)
U8                           hinting                      0 - none, 1 - light, 2 - full (ignored for bitmap fonts)
U8                           msdf                         0 - non MSDF, 1 - MSDF
U8                           autohinter                   0 - use font built-in hinting, 1 - use auto hinter (ignored for bitmap fonts)
U8                           antialiased                  0 - 1-bit rendering, 1 - 8-bit rendering (ignored for MSDF and bitmap fonts)
U8                           spacing_glyph                extra glyph spacing in pixels
U8                           spacing_space                extra space spacing in pixels
F32                          oversampling                 oversampling factor (if <= 0 - global value is used, ignored for MSDF and bitmap fonts)
U32                          lang_support_overrides       number of language support overrides
"lang_support_overrides" records:
    U32                      lang_size                    length of language code string
    U8[lang_size]            lang_code                    language ISO code as UTF-8 string
    U8                       lang_enable                  0 - language not supported, 1 - language supported
U32                          script_support_overrides     number of script support overrides
"script_support_overrides" records:
    U32                      script_size
    U8[script_size]          script_code                  script ISO code as UTF-8 string
    U8                       script_enable                0 - script not supported, 1 - script supported
F32                          rect_margin                  extra spacing between glyphs in texture (ignored for bitmap fonts)
F32                          px_range                     MSDF pixel range (ignored for non MSDF fonts)
U32                          base_variation_tags          number of variation tags in base variation
"base_variation_tags" records:
    U32                      variation_tag                OpenType variation tag
    F64                      variation_value              value for variation tag
U16                          base_size                    base font size
U32                          cached_variations            number of cached font variants (always 1 for bitmap fonts)
"cached_variations" records:
    U32                      variation_tags               number of variation tags in this variation
    "variation_tags" records:
        U32                  variation_tag                OpenType variation tag
        F64                  variation_value              value for variation tag
    U32                      cached_sizes                 number of cached font sizes for this variation (always 1 for bitmap and MSDF fonts)
    "cached_sizes" records:
        U32                  size                         font size (always equal to base_size for bitmap fonts)
        F32                  ascent                       font ascent
        F32                  descent                      font descent
        F32                  height                       font height
        F32                  underline_position           font underline position
        F32                  underline_thickness          font underline thickness
        U32                  textures                     number of textures
        "textures" records:
            U32              width                        texture width
            U32              height                       texture height
            U32              offsets                      number of texture offsets
            U16[offsets]     offset_value                 texture offset values
            U32              name_size
            U8[name_size]    texture_name                 name of texture file
        U32                  glyphs                       number of cached glyphs
        "glyph" records:
            U32              index                        glyph index (for dynamic fonts) or UTF-32 code (for bitmap fonts)
            U8               found                        0 - glyph is not in the font (rest of the glyph data ignored) , 1 - glyph is in the font
            U16              texture_idx                  index of texture in the "textures" array
            F32              rect_pos_x
            F32              rect_pos_y                   glyph offset from the baseline
            F32              rect_size_x
            F32              rect_size_x                  glyph size
            F32              uv_pos_x
            F32              uv_pos_y
            F32              uv_size_x
            F32              uv_size_x                    glyph position and size in the texture
            F32              advance_x                    glyph advance for horizontal layout
            F32              advance_y                    glyph advance for vertical layout
        U32                  kerning_size                 number of kerning pairs (always 0 for dynamic fonts)
        "kerning_size" records:
            U32              glyph_index_a                first character UTF-32 code
            U32              glyph_index_b                second character UTF-32 code
            U32              value                        kerning offset value
==============================================================================================================================================
```

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/2021 and https://github.com/godotengine/godot/issues/28045 (by implementing per-font oversampling override).*

bruvzg committed:
Font M(T)SDF and import support.

Add multi-channel + true SDF font texture generation and rendering support.
Add FontData import plugin, font texture cache pre-generation and loading, move font properties from resource to imported cache.
Add bitmap font generation support.

Calinou commented:
This looks great!

I noticed the MSDF becomes quite "thin" at lower zoom values. Is it possible to force using a minimal amount of sharpness/width for the MSDF font at the cost of worse antialiasing? This would help keep the font more readable at lower font sizes.

bruvzg commented:
> I noticed the MSDF becomes quite "thin" at lower zoom values. Is it possible to force using a minimal amount of sharpness/width for the MSDF font at the cost of worse antialiasing?

Probably, I have not tested any SDF generation parameters yet, and used the default shader with AA as is.

fire commented:
I wonder how this approach compares? https://github.com/hypernewbie/VEFontCache

summertimejordi commented:
Is this for the master branch or 3.2?

Calinou commented:
@summertimejordi This is for the `master` branch only, not 3.2. It can't be backported due to the reliance on TextServer changes;

bruvzg commented:
_Update:_

Contour orientation detection and alignment issues are fixed, self intersection elimination is still desired, but rare and usually cause only small artifacts, probably can be omitted for now. Seems to work fine with all fonts I have tested.

<details>
  <summary>Screenshots - click to expand</summary>

<img width="57" alt="Screenshot 2021-02-23 at 11 55 45" src="https://user-images.githubusercontent.com/7645683/108830172-bc23a780-75d1-11eb-8c8a-bbe807b9e3fd.png">
<img width="362" alt="Screenshot 2021-02-23 at 11 55 59" src="https://user-images.githubusercontent.com/7645683/108830175-bded6b00-75d1-11eb-8982-54b6e85c8e14.png">
<img width="1103" alt="Screenshot 2021-02-23 at 11 56 15" src="https://user-images.githubusercontent.com/7645683/108830187-bfb72e80-75d1-11eb-8f9d-7a6c7ae06fd6.png">

</details>

Stuff still missing (help needed):
- Outline drawing shader.
- Some AA probably should be added to the shader, small sizes are too sharp.
Current shader is doing absolute minimum:
https://github.com/godotengine/godot/blob/490e3108c9618636d84c08a56b155ad6841c78ec/servers/rendering/renderer_rd/shaders/canvas.glsl#L457-L479

Other stuff:
- Import progress dialog is not displayed, am I missing something, or it's editor bug?
- I'm not sure what's the best way to store cached font textures (right now they are store as `png`s alongside `fontdata`).
<img width="562" alt="Screenshot 2021-02-23 at 12 30 59" src="https://user-images.githubusercontent.com/7645683/108831585-0194a480-75d3-11eb-8523-73410b6fd330.png">

bruvzg commented:
Also, exposed M(T)SDF drawing function to the canvas item for custom textures (`draw_msdf_texture_rect_region`).

<img width="391" alt="Screenshot 2021-02-23 at 16 19 46" src="https://user-images.githubusercontent.com/7645683/108856810-1aad4d80-75f3-11eb-8fee-8a7e4d04141e.png">

Image taken from [https://github.com/V-Sekai/godot-msdf-project/](https://github.com/V-Sekai/godot-msdf-project/) (`output.png`).

bruvzg commented:
msdfgen seems to have issues with sharp internal angles, usually resulting artifacts are small and barely noticeable. Unfortunately outlines generated by [FreeType glyph stroker](https://www.freetype.org/freetype2/docs/reference/ft2-glyph_stroker.html) have tons of internal angles (in addition of common self-intersections) and the resulting distance field is total mess.

Distance filed for outline:
![Estedad-VF ttf-0962a39dac2b74b3641896668f7ce29e 1505 50080 0](https://user-images.githubusercontent.com/7645683/109011006-22372a00-76b9-11eb-800c-bd8c21ca8747.png)

Small artifacts on the normal glyph (zoomed in to 3500%):
<img width="306" alt="Screenshot 2021-02-24 at 15 25 56" src="https://user-images.githubusercontent.com/7645683/109011137-44c94300-76b9-11eb-82d5-f48c7be09811.png">

On the other hand, simply applying radial blur to the "true distance" component of M(T)SDF seems to have almost the same effect. Probably it's more than enough for all real use cases of the outline.

<img width="1767" alt="Screenshot 2021-02-24 at 14 52 46" src="https://user-images.githubusercontent.com/7645683/109011680-d769e200-76b9-11eb-89ea-4b838d349604.png">

MSDF font on the top buttons, dynamic font on bottom buttons, scaled to: 25%, 50%, 75%, 100%, 200% and 500%.

bruvzg commented:
I have changed outline rendering from "blur" to using max value in range, now MSDF font outlines look the same as dynamic font outlines at all sizes (right now MSDF font outline texture size is limited to 3 x glyph size, which should be more than enough for any real use case).

Also, added API for individual font oversampling override (which was used to disable oversampling for MSDF fonts).

<img width="1761" alt="Screenshot 2021-02-26 at 10 17 54" src="https://user-images.githubusercontent.com/7645683/109274696-79a1db00-781c-11eb-8f86-c90460a5aca5.png">

Everything seems to be working as expected. I'll do a bit more testing and performance profiling, and it should be ready to go.

fire commented:
Is it possible to render MSDF text directly to a Texture2D so it can be used in a 3d sprite or other uses?

Calinou commented:
@fire You can do so using a Viewport (as you would do currently).

bruvzg commented:
#### Possible future improvements:

- MSDF generation can be parallelized, but msdfgen is using OpenMP, which was one rejected for use in Godot, and therefore is disabled. But it's only a few for loops, probably worth modifying it using to use Godot own threading.
- Replacing default export template bitmap HiDPI/LoDPI fonts with MSDF bitmap font seems like a good idea.
- Replacing editor fonts with MSDF fonts might be useful for fully scalable editor implementation, but it involves pre-rendering a large number of glyphs for all locales. To avoid bloating editor, pre-rendering for the specific language can be done on the user machine when it's first selected or be a part of the language pack plugin (along with a translation file).
- The same can be done for the editor SVG icons (some extra work needed to support multicolor icons).

#### Some import related stuff that do not work, but is not specific to this PR (same is true for all other imported resources):

- Reimporting do not reload fonts used in the scene.
- Import progress bar is only visible in single-window mode.
- Reimporting do not clean old import files (e.g. if, after reimport, used texture number is smaller, old file will stay).

bruvzg commented:
> Is it possible to render MSDF text directly to a Texture2D so it can be used in a 3d sprite or other uses?

Is there any standard way to render anything directly to texture other than using viewports? Viewports should work fine, and I do not think adding some new way to do it only for the text is worth it.

fire commented:
I tried the pr and create a rich text field. I zoomed in and the MSDF wasn't blurry. [English Alphabet] compared to the regular version. 

The spacing changed. That's strange but acceptable.

fire reviewed (approved):
I briefly looked at the code, earlier I tried the basic use of the feature.
The MSDF pr looks good from a feature standpoint. I tried it and it seemed to do what it said it did.

Future possible things: 
1. Performance review on mobile devices / laptop integrated gpus.
2. Possibility to output the msdf raw rgba values (so that the viewport use can be done and we can pass it through to a compositing shader).

bruvzg commented:
> Performance review on mobile devices / laptop integrated gpus.

Rendering performance should be almost the same as for normal dynamic fonts (a bit higher for outlines), it's basically sample, take median value of all channels and do some range adjustment. Generation is done on CPU.

> Possibility to output the msdf raw rgba values (so that the viewport use can be done and we can pass it through to a compositing shader).

It's easy to do but, glyph textures usually have overlapping parts. I'm not sure how well blending multiple raw MSDF textures over each other will work, is the result of such process still a valid distance field, or we need some special kind of blending?

lyuma commented:
> It's easy to do but, glyph textures usually have overlapping parts. I'm not sure how well blending multiple raw MSDF textures over each other will work, is the result of such process still a valid distance field, or we need some special kind of blending?

I believe SDF are designed to be blendable with a simple "minimum" operation.

I don't know if it will work ideally in the case that glyphs overlap or intersect, but it should produce some reasonable output.... if it's causing artifacts, I imagine they can be mitigated by increasing the viewport resolution.

In case you weren't aware, it is possible to tell the GPU to use a minimum blending operation instead of the normal "mix" style blending we are used to. In terms of the actual graphics APIs, this should be widely supported (even GLES2 has an extension): In OpenGL, that would be `glBlendEquation(GL_MIN)` and In Vulkan, it is `VKBlendOp` with `VK_BLEND_OP_MIN`.
https://www.khronos.org/registry/vulkan/specs/1.0/html/chap28.html#framebuffer-blendoperations

bruvzg commented:
Added function to disable distance field processing shader for the `FontData` instance (not included to import options, since it's not changing anything in imported data). Blending mode can be set via canvas material properties (`min` seems to be correct one).

jordo commented on servers/rendering/renderer_rd/shaders/canvas.glsl:
```diff
@@ -449,7 +457,54 @@ void main() {
+
+			float a = 0.0;
+			float r2 = outline_thickness * outline_thickness / 4;
+			for (int y = 0; y < int(outline_thickness); y++) {
```
With an sdf can't outline be done in a single lookup as well?   I can't remember the details of msdf off the top of my head, but I recall doing outline in a single lookup in the past.  Also should be able to get perfect antialiasing on the outline with fwidth the same as without an outline.

Calinou replied:
Indeed, supporting soft outlines would be useful for things like soft shadows. You can already do those using shaders, but a built-in way will be more convenient and possibly faster.

bruvzg replied:
> With an sdf can't outline be done in a single lookup as well?

It is possible, but work well only for thin outlines. Maybe I'm missing something. And `FTStroker` produce self-intersections and sharp angles, causing msdfgen generation to break.

A better outline shader would be appreciated, if you have any ideas, give it a try.

> With an sdf, it would even be nice to add a 'falloff' uniform as well as thickness... Quite useful for some effects.

Existing shader is reusing existing nine patch uniform for `pixel range` and `outline thickness` which have two unused members, should be possible to add it without any extra cost.

jordo replied:
> It is possible, but work well only for thin outlines. Maybe I'm missing something. And `FTStroker` produce self-intersections and sharp angles, causing msdfgen generation to break.
> 
> A better outline shader would be appreciated, if you have any ideas, give it a try.

See last two posted screens for a thick SDF outline.  Once you have your geometry defined as a distance field you don't need to muck around with freetype stroker at all (shouldn't be used at all).  SDF will be more efficient and is pretty performant for doing this stuff.  Single lookup in fragment shader.

jordo commented:
@bruvzg Ya, you can easily do the outline in a single texture lookup if you use the SDF there as well.  See below screenshots:

**Current outline:**
![Screen Shot 2021-03-01 at 11 35 01 AM](https://user-images.githubusercontent.com/323868/109535551-511b1a80-7a82-11eb-893f-c5e189468bd6.png)

**Current outline zoomed in (two for loops and multi-sampled):**
![Screen Shot 2021-03-01 at 11 35 22 AM](https://user-images.githubusercontent.com/323868/109535564-56786500-7a82-11eb-8cec-5bd0e6d7b294.png)


**Single SDF lookup:**
![Screen Shot 2021-03-01 at 11 32 46 AM](https://user-images.githubusercontent.com/323868/109535596-5f693680-7a82-11eb-94d6-efa880d701ae.png)

**Single SDF lookup zoomed in:**
![Screen Shot 2021-03-01 at 11 32 53 AM](https://user-images.githubusercontent.com/323868/109535614-63955400-7a82-11eb-9818-94e465933e68.png)

bruvzg commented:
> @bruvzg Ya, you can easily do the outline in a single texture lookup if you use the SDF there as well. See below screenshots:

Is it working with outline size > 1, anything above 10 is probably useless but 5-7 might be.
<img width="253" alt="Screenshot 2021-03-01 at 19 41 53" src="https://user-images.githubusercontent.com/7645683/109536356-4da98280-7ac6-11eb-9845-c6d5f7a5d488.png">

jordo commented:
OK, I can modify the outline glsl after the PR is accepted or whatever really.  Works great as long as within whatever size the backing outline texture has.  It seems the backing texture doesn't dynamically scale or resize itself.

But also, some general thoughts:
 - Outline for an msdf label doesn't need to be done in another quad...  It can easily be done in one pass, it's probably more efficient to do so.  No need to have multiple.  Actually all the effects that normally require multi-sampling (shadow, etc), can be done in a single lookup in fragment shader with an SDF.  

 - Not sure if this is specific to msdf or not, but there's a lot of overdraw happening on fonts that aren't even using an outline:  See below video:   This is what is getting drawn regardless if outline is even checked or not.  If outline isn't checked it just rendering transparent pixels, which is really wasteful.  


https://user-images.githubusercontent.com/323868/109536608-8d02af80-7a83-11eb-9989-3bc8a66c03c0.mov

bruvzg commented:
> It seems the backing texture doesn't dynamically scale or resize itself.

Outline textures hardcoded to 3x of glyph texture.

> Outline for an msdf label doesn't need to be done in another quad... It can easily be done in one pass, it's probably more efficient to do so. No need to have multiple. Actually all the effects that normally require multi-sampling (shadow, etc), can be done in a single lookup in fragment shader with an SDF.

Single pass won't work with overlapping glyphs:
<img width="515" alt="Screenshot 2021-03-01 at 19 55 39" src="https://user-images.githubusercontent.com/7645683/109537951-31a6e080-7ac8-11eb-822a-190e7d11e7bc.png">

> Not sure if this is specific to msdf or not, but there's a lot of overdraw happening on fonts that aren't even using an outline: See below video: This is what is getting drawn regardless if outline is even checked or not. If outline isn't checked it just rendering transparent pixels, which is really wasteful.

Default outline color is (1,1,1,1), unchecking outline color won't disable it (only setting outline size to 0 do).

jordo commented:
> Outline textures hardcoded to 3x of glyph texture.

Ah, I see.  In that case perhaps the default distance field generation for msdf should also be 3x glyph texture to support outlines out to the maximum outline texture.

> Single pass won't work with overlapping glyphs:

Yep, forgot they are single glyphs.  

> Default outline color is (1,1,1,1), unchecking outline color won't disable it (only setting outline size to 0 do).

OOOO! nice that's good, but a total gotcha.  (Not intutive at least).  At any rate, good stuff.  Can't wait for this to be merged.

jordo commented:
@bruvzg  attached video of a thick sdf outline.  This is a single texture lookup to do outlines and effects like this with an sdf.  Extremely fast.  The supported thickness is going to be whatever range is in your generated field is.  I'd almost just suggest we simplify the range (perhaps not even expose it as configurable).  If say outline is already hard coded to 3x size, then perhaps field generation should just line up 1-1.  

https://user-images.githubusercontent.com/323868/109542489-bbd05400-7a8a-11eb-9360-cbc742c29de6.mov

jordo commented:
Arg, video compression is really hard to tell difference.  Two screens below:

Zoomed in current thick outline.  (Filter based and very expensive in the fragment shader):  
<img width="1568" alt="Screen Shot 2021-03-01 at 12 51 07 PM" src="https://user-images.githubusercontent.com/323868/109544199-e1f6f380-7a8c-11eb-8733-c6242e6f0425.png">

Zoomed in with an SDF outline. (A single texture read in the fragment):
<img width="1742" alt="Screen Shot 2021-03-01 at 12 46 24 PM" src="https://user-images.githubusercontent.com/323868/109544313-02bf4900-7a8d-11eb-94a6-1b4a26e74870.png">

bruvzg commented:
Added option to set MSDF resolution:
- Uses same base variation size value:
<img width="409" alt="Screenshot 2021-03-01 at 23 09 19" src="https://user-images.githubusercontent.com/7645683/109559394-46dd3880-7ae3-11eb-866d-3857f8a8a994.png">

- All previous screenshots are rendered with resolution 128 and pixel range 10.
- Smaller resolution require smaller pixel range.

Resolution 32, pixel range 4 example:
<img width="268" alt="Screenshot 2021-03-01 at 23 09 46" src="https://user-images.githubusercontent.com/7645683/109559514-6d02d880-7ae3-11eb-9d71-3f75f0f99fec.png">

jordo commented:
@bruvzg Great work on this!  I feel like this will be my goto font rendering mode in the engine.  Really looking forward to using it.

That last push maybe broke defaults (At least for my test font, and default label size).  (Is default 16 now?)  Maybe base variation should be larger be default?

![Screen Shot 2021-03-01 at 3 42 57 PM](https://user-images.githubusercontent.com/323868/109563011-cdbef080-7aa4-11eb-8db3-8aad8c343682.png)

Should we really expose both the msdf field range and the base variation size?  Could we just choose the field size to be something that works for 99% of use cases?  I feel like the msdf field range could be derived from the base size, and in a way that fits almost all use cases in engine (3x max outline and shadow size for example).   I feel like that would be a sane default, and probably work well most of the time.  

In general these msdf work really great even at small sizes so I feel like we could default to something pretty liberal? 32 or even 64 per glyph is still pretty decent quality for size?

bruvzg commented:
- Added import presets (MSDF font → size = 64, Dynamic font → size = 16).
- Removed hard-coded 3x texture size, texture should have extra `pixel_range` thick border on all sides, which should be sufficient for any supported outline sizes.
- Changed outline rendering to single sample shader, outline size is clamped to `pixel_range / 2`.
- Also decreased non MSDF strocker based outline scaling to allow thin outlines.

New outlines of size 2 and 12:
<img width="1016" alt="Screenshot 2021-03-02 at 10 09 05" src="https://user-images.githubusercontent.com/7645683/109617698-630ec300-7b3f-11eb-8602-17ca5da2cbd8.png">

jordo commented:
@bruvzg Awesome work, just perfect!!!  Can you hold off force pushing after review starts?  I'm having a hard time tracking any delta change with every force push.

bruvzg committed:
Add `MSDFLoader` class to convert `SVG`s and arbitrary vector shapes to distance fields.

bruvzg commented:
Rebased and added  `MSDFLoader` class to convert `SVG`s or/and arbitrary vector shapes to distance fields (as second commit, there are no changes in the first one).

Distance field image size is determined from the vector shape bounds, SVG can be scale with the second `load_svg`argument.

<img width="1358" alt="Screenshot 2021-03-04 at 13 57 55" src="https://user-images.githubusercontent.com/7645683/109960584-a7da5b80-7cf1-11eb-854a-a4fdd0826a8c.png">

SDF image size on the screenshot is `76` x `76`.

fire commented:
I think it's reasonable to bind "*.svg" to the svg to msdf loader as the second non-default slot, but no big preference.

bruvzg commented:
Rebased and integrated bitmap font scaling fix (#46782), no changes to MSDF.

bruvzg committed:
Implement advanced import settings dialog for the FontData.

bruvzg commented:
Rebased and added advanced import settings dialog for more convenient preloaded glyph selection (as separate commit):

<img width="1465" alt="font_import" src="https://user-images.githubusercontent.com/7645683/112303480-010c2e00-8ca5-11eb-9216-0fb5d6590738.png">

Calinou commented:
@bruvzg On the glyph selection window, it seems the headings don't use the bold font as they should. Instead, they use the default project bitmap font.

bruvzg commented:
> On the glyph selection window, it seems the headings don't use the bold editor font as they should. Instead, they use the default project bitmap font.

Moved label font init to the `NOTIFICATION_ENTER_TREE`.

> Also, for the hexadecimal ranges, I think it'd be better to use lowercase characters for letters as it makes it easier to distinguish between 8 and b. We already try to use lowercase for hexadecimal numbers in other places in Godot.

I do not have any preference, but since most hex uses in the editor seems to use low case, I have changed it to low case as well.

<img width="945" alt="Screenshot 2021-03-24 at 16 39 52" src="https://user-images.githubusercontent.com/7645683/112329552-f9a64e00-8cbf-11eb-88ef-ecaeeac4c0e1.png">

reduz commented:
So far this looks good to me for the most part. I really dislike a custom binary format, though. Can we use regular resource serializing here?

fire commented:
There may be some bug fixes for corner bugs. See https://github.com/Chlumsky/msdfgen/releases/tag/v1.9.1

bruvzg commented:
Superseded by #51908
